include/xtensor-r/rcpp_extensions.hpp

```diff
@@ -0,0 +1,108 @@
+/***************************************************************************
+* Copyright (c) 2017, Wolf Vollprecht, Johan Mabille, and Sylvain Corlay   *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_R_RCPP_EXTENSIONS_HPP
+#define XTENSOR_R_RCPP_EXTENSIONS_HPP
+
+
+#include <Rcpp.h>
+
+namespace xt
+{
+    template <class T>
+    class rarray;
+
+    template <class T, std::size_t N>
+    class rtensor;
+}
+
+namespace Rcpp
+{
+    template <class T>
+    class ReferenceInputParameter<xt::rarray<T>>
+    {
+    public:
+        typedef xt::rarray<T>& reference ;
+        ReferenceInputParameter(SEXP x_)
+            : obj(x_)
```

Comment:
Looking again at this, you should not need to specialize this at all.

In the original

```
ReferenceInputParameter(SEXP x_) : obj( as<T>(x_) ){}
```

`as<T>(x_)` returns an rvalue, so the move constructor of `T` should be called and not the copy constructor.